### PR TITLE
Skip update if download failed before

### PR DESCRIPTION
**Pull request by mike-sul**
aklite: Skip update download if failed before
    
Skip an attempt to download an update/Target if:
    1) its ostree download failed before due to lack of free space;
    2) size of free space has not increased since the failure time.

### CODE REVIEW

**Comment by mike-sul** [comment]
@doanac I tried to implement it in truly cowboy way :).

File: src/main.cc [hunk 1]
```diff
@@ -272,6 +281,8 @@ static int daemon_main(LiteClient& client, const bpo::variables_map& variables_m
 
   client.reportAktualizrConfiguration();
 
+  std::tuple<Hash, boost::uintmax_t, std::string> available_space_when_download_failed{Hash{"", ""}, 0, ""};
```
Comment (doanac) [inline src/main.cc:284]: trying to think through what happens here for a reboot. The path, `available_space_when_download_failed.2` won't be set, so the check for disk space will fail and we'll try and do an install. 

I *think* this is the right approach, just confirming i'm reading this right?

Comment (mike-sul) [reply src/main.cc:284]: Yes, that's correct. I am just not sure it's worth persisting the error state/size.

File: src/main.cc [hunk 2]
```diff
@@ -349,11 +360,35 @@ static int daemon_main(LiteClient& client, const bpo::variables_map& variables_m
...
+              std::get<0>(available_space_res) <
+                  (std::get<1>(available_space_when_download_failed) + 1024 /* at least 1K should be freed */)) {
+            LOG_ERROR << "Free space has not been increased since the failure to download Target's ostree happened; "
+                         "ostree hash: "
+                      << target_to_install.sha256Hash()
+                      << ", free space when download failed: " << std::get<1>(available_space_when_download_failed)
+                      << ", current free space: " << std::get<0>(available_space_res);
+
+            // Do we need to inform backend about it? If yes then how, recreate the original flow, downloadstart,
+            // downloadfailed...
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/main.cc:379]: Let me check whether the correlation ID set to Target instance will be the same across download attempts.
If so, then we can keep sending the download events. Should we send downloadstart and downloadcomplete pair? Do we need to send the original "no space" error of this one "skip download"?

File: src/main.cc (hunk 1 above)
Comment (doanac) [reply src/main.cc:284]: i don't think so. just makes things complicated. Plus, reboots fix everything :)

File: src/main.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (doanac) [reply src/main.cc:379]: i *think* libaktualizr may keep it around in sqlite. I think we can just do a downloadComplete with failed status and maybe some disk space info

[the remark this answers is not in the document]
Comment (mike-sul) [reply src/main.cc]: No, the libaklite does not store the modified Target (with the set cor ID) if its update fails, so it really is available in the liteclient context.

Nevertheless, I managed to do it, see the latest version. The update logs on the backend will look like the following:

```
fioctl device updates 6e463bfe-6ffa-42ff-af23-8cee513ad9b7 1312-76a5173b-b9fb-42db-b941-ea5267c1bdfd
2022-05-03T09:50:23+00:00 : EcuDownloadStarted(intel-corei7-64-lmp-1312)
 Details:
 | Updating from unknown to intel-corei7-64-lmp-1312
2022-05-03T09:50:25+00:00 : EcuDownloadCompleted(intel-corei7-64-lmp-1312) -> Failed!
 Details:
 | Insufficient storage available; path: /work/fio/prjs/lmp-device-register/device/sysroot; err: Writing content object: min-free-space-size 1048576MB would be exceeded, at least 4.1 kB requested
2022-05-03T09:50:33+00:00 : EcuDownloadCompleted(intel-corei7-64-lmp-1312) -> Failed!
 Details:
 | No space to download Target's ostree; hash: df89f611a70b20ddffba6c264705b4821fea2d521f653369f1f831759c89f46b, required: 413455655936, available: 413455654912
2022-05-03T09:50:42+00:00 : EcuDownloadCompleted(intel-corei7-64-lmp-1312) -> Failed!
 Details:
 | No space to download Target's ostree; hash: df89f611a70b20ddffba6c264705b4821fea2d521f653369f1f831759c89f46b, required: 413455655936, available: 413455654912

```

[the remark this answers is not in the document]
Comment (doanac) [reply src/main.cc]: nice work. looks good

**Pull request merged**